### PR TITLE
Add region information in :serverinfo

### PR DESCRIPTION
**Pull request by LolloDev5123**
Roblox 10_07_2021 05_48_20 <-- got inspired by arsenal kiddies
random stuff i make for adonis - Roblox Studio 10_07_2021 06_35_05

**-- IMPORTANT --**
I do not know if it would be ethical to include an IP Address, it is very possible. I have fear this may be exploitable and make it easier for skids to DDOS servers (already possible)
Second issue: When in studio, the ip could reveal yours to unwanted people, so I don't know if it is good to add one and if we do, we might censor it in studio.

Please note that there may be some bugs, such as error handling, checking if HTTP is on, but on the other hand, this would be a pretty good addition to :serverinfo or :serverdetails.

### CODE REVIEW

**Comment by LolloDev5123** [comment]
do note the region and things are approximations so yeah while i do live in italy im not even in apulia or veglie lol

File: MainModule/Server/Commands/Moderators.lua [hunk 1]
```diff
@@ -1656,13 +1656,40 @@ return function(Vargs, env)
...
+					data.city = city --City: Veglie
+					data.region = region--Region: Apulia
+					data.zipcode = zipcode--Zipcode: 73010
+					data.timezone = timezone--Timezone: Europe/Rome
+					-- Server IP Address is not grabbed for ethical purposes.
+					-- Seriously tho do optimize this code k thx
+                                        
+					return data
+				end
+				local serii = getServerInternetInfo()
```
Comment (joritochip) [inline MainModule/Server/Commands/Moderators.lua:1684]:
```suggestion
				local s, r = pcall(service.HttpService.GetAsync, service.HttpService, "http://ip-api.com/json")
				if s then
					r = service.HttpService:JSONDecode(r)
				end
				local serverInfo = s and {
					country = r.country 
					city = r.city 
					region = r.region 
					zipcode = r.zipcode 
					timezone = r.timezone 
				} or nil
```

File: MainModule/Server/Commands/Moderators.lua [hunk 2]
```diff
@@ -1656,13 +1656,40 @@ return function(Vargs, env)
...
 				Remote.MakeGui(plr,"ServerDetails",{
 					CreatorId = game.CreatorId;
 					PrivateServerId = game.PrivateServerId;
 					PrivateServerOwnerId = game.PrivateServerOwnerId;
 					ServerStartTime = service.FormatTime(server.ServerStartTime);
 					ServerAge = service.FormatTime(os.time()-server.ServerStartTime);
 					HttpEnabled = HTTP.CheckHttp();
+					ServerInternetInfo = serii;
```
Comment (joritochip) [inline MainModule/Server/Commands/Moderators.lua:1692]:
```suggestion
					ServerInternetInfo = serverInfo;
```

File: MainModule/Client/UI/Default/ServerDetails.lua [hunk 3]
```diff
@@ -100,6 +105,14 @@ return function(data)
...
 			addOverviewEntry("Private Server Owner:", (game:GetService("Players"):GetNameFromUserIdAsync(data.PrivateServerOwnerId) or "[Unknown Username]").." ("..game.PrivateServerOwnerId..")")
 		end
+		--Server Internet Info
+		serii = data.ServerInternetInfo
+		addOverviewEntry("Timezone:", serii.timezone or "[Error]")
+		addOverviewEntry("Country:", serii.country or "[Error]")
+		addOverviewEntry("Region:", serii.region or "[Error]")
+		addOverviewEntry("City:", serii.city or "[Error]")
+		addOverviewEntry("Zipcode:", serii.zipcode or "[Error]")
+
```
Comment (joritochip) [inline MainModule/Client/UI/Default/ServerDetails.lua:115]:
```suggestion
		if data.ServerInternetInfo then
			--Server Internet Info
			local serii = data.ServerInternetInfo
			addOverviewEntry("Timezone:", serii.timezone or "[Error]")
			addOverviewEntry("Country:", serii.country or "[Error]")
			addOverviewEntry("Region:", serii.region or "[Error]")
			addOverviewEntry("City:", serii.city or "[Error]")
			addOverviewEntry("Zipcode:", serii.zipcode or "[Error]")
		end
```

**Review by joritochip — CHANGES_REQUESTED** [review]

File: MainModule/Server/Commands/Moderators.lua [hunk 4]
```diff
@@ -1656,13 +1656,25 @@ return function(Vargs, env)
...
+				local s, r = pcall(service.HttpService.GetAsync, service.HttpService, "http://ip-api.com/json")
+				if s then
+					r = service.HttpService:JSONDecode(r)
+				end
+				local serverInfo = s and {
+					country = r.country 
+					city = r.city 
+					region = r.region 
+					zipcode = r.zipcode 
+					timezone = r.timezone
```
Comment (joritochip) [inline MainModule/Server/Commands/Moderators.lua:1668]:
```suggestion
					country = r.country,
					city = r.city,
					region = r.region,
					zipcode = r.zip,
					timezone = r.timezone
```

**Review by joritochip — CHANGES_REQUESTED** [review]
oops

**Comment by P3tray** [comment]
Sir just add an :IsStudio and if true remove the location info from the list. It also means you can include the IP then.

For some of us leaking our city is quite bad considering their size in rural countries. My home location doesn't exactly have millions living there.

IsStudio would fix that.

**Comment by Kw6m** [comment]
I don't think you can get the players IP... Only the servers IP. If I am wrong do point it out to me please. But if the IP is available to the server. Don't display it. Region sure. But IP? The neanderthal's that would be interested in peoples IP's wouldnt be using it for good.

**Comment by P3tray** [comment]
> I don't think you can get the players IP... Only the servers IP. If I am wrong do point it out to me please. But if the IP is available to the server. Don't display it. Region sure. But IP? The neanderthal's that would be interested in peoples IP's wouldnt be using it for good.
> 
> But yes showing the servers IP does no harm. Showing the servers location _also_ does no harm

This is the IP for the server. If this is a studio test server than it'll be ran from your pc, and therefore your IP.

Not good for taking screenshots or working in teamcreate "anonymously".

Hence my suggestion.

And technically speaking you can't take IP's from clients.

**Comment by Sceleratis** [comment]
I see no harm in showing the server's IP.

**Comment by LolloDev5123** [comment]
It could lead to DDOSING, but hey, an exploiter with a potent enough exploit could already get the IP, so, I can show the IP and coordinates? The city is always an approximation, the only things that are right are the region, country, and timezone, I would personally blur out city, IP, and region when someone is in studio, as you could give too much details when screensharing a playtest.

**Comment by coasterteam** [comment]
> I see no harm in showing the server's IP.

As long as it's only shown to people with elevated permissions

**Comment by Sceleratis** [comment]
> 
> 
> It could lead to DDOSING, but hey, an exploiter with a potent enough exploit could already get the IP, so, I can show the IP and coordinates? The city is always an approximation, the only things that are right are the region, country, and timezone, I would personally blur out city, IP, and region when someone is in studio, as you could give too much details when screensharing a playtest.

Um. You can already see the IP of the server you're connected to... no exploits required. So....

**Comment by joritochip** [comment]
> > It could lead to DDOSING, but hey, an exploiter with a potent enough exploit could already get the IP, so, I can show the IP and coordinates? The city is always an approximation, the only things that are right are the region, country, and timezone, I would personally blur out city, IP, and region when someone is in studio, as you could give too much details when screensharing a playtest.
> 
> Um. You can already see the IP of the server you're connected to... no exploits required. So....
> 
> Also as to the studio thing I feel like that's kinda niche and would be their own fault, but if it really bothers you, you can just check if they're in studio or not and then return something like "N/A" or "Studio"

It is also possible to get the IP of any given server without even connecting to it through the API, so I would agree in saying that having the IP there is not that big of an issue.

**Comment by LolloDev5123** [comment]
Alright, I guess this is ready to be pushed? Senstive data such as Region, IP, City, and ZIP Code are not shown when play testing.

**Comment by LolloDev5123** [comment]
Check this stuff for the last time as this is good to be merged I guess

**Pull request merged**